lightning/src/offers/offer.rs

```diff
@@ -706,4 +831,266 @@ mod tests {
 		assert_eq!(offer.supported_quantity(), Quantity::one());
 		assert_eq!(tlv_stream.quantity_max, None);
 	}
+
+	#[test]
+	fn parses_offer_with_chains() {
+		let offer = OfferBuilder::new("foo".into(), pubkey(42))
+			.chain(Network::Bitcoin)
+			.chain(Network::Testnet)
+			.build()
+			.unwrap();
+		if let Err(e) = offer.to_string().parse::<Offer>() {
+			panic!("error parsing offer: {:?}", e);
+		}
+	}
+
+	#[test]
+	fn parses_offer_with_amount() {
+		let offer = OfferBuilder::new("foo".into(), pubkey(42))
+			.amount(Amount::Bitcoin { amount_msats: 1000 })
+			.build()
+			.unwrap();
+		if let Err(e) = offer.to_string().parse::<Offer>() {
+			panic!("error parsing offer: {:?}", e);
+		}
+
+		let mut tlv_stream = offer.as_tlv_stream();
+		tlv_stream.amount = Some(1000);
+		tlv_stream.currency = Some(b"USD");
+
+		let mut encoded_offer = Vec::new();
+		tlv_stream.write(&mut encoded_offer).unwrap();
+
+		if let Err(e) = Offer::try_from(encoded_offer) {
+			panic!("error parsing offer: {:?}", e);
+		}
+
+		let mut tlv_stream = offer.as_tlv_stream();
+		tlv_stream.amount = None;
+		tlv_stream.currency = Some(b"USD");
+
+		let mut encoded_offer = Vec::new();
+		tlv_stream.write(&mut encoded_offer).unwrap();
+
+		match Offer::try_from(encoded_offer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingAmount)),
+		}
+	}
+
+	#[test]
+	fn parses_offer_with_description() {
+		let offer = OfferBuilder::new("foo".into(), pubkey(42)).build().unwrap();
+		if let Err(e) = offer.to_string().parse::<Offer>() {
+			panic!("error parsing offer: {:?}", e);
+		}
+
+		let mut tlv_stream = offer.as_tlv_stream();
+		tlv_stream.description = None;
+
+		let mut encoded_offer = Vec::new();
+		tlv_stream.write(&mut encoded_offer).unwrap();
+
+		match Offer::try_from(encoded_offer) {
+			Ok(_) => panic!("expected error"),
+			Err(e) => {
+				assert_eq!(e, ParseError::InvalidSemantics(SemanticError::MissingDescription));
+			},
+		}
+	}
+
+	#[test]
+	fn parses_offer_with_paths() {
+		let offer = OfferBuilder::new("foo".into(), pubkey(42))
+			.path(BlindedPath {
+				introduction_node_id: pubkey(40),
+				blinding_point: pubkey(41),
+				blinded_hops: vec![
+					BlindedHop { blinded_node_id: pubkey(43), encrypted_payload: vec![0; 43] },
+					BlindedHop { blinded_node_id: pubkey(44), encrypted_payload: vec![0; 44] },
+				],
+			})
+			.path(BlindedPath {
+				introduction_node_id: pubkey(40),
+				blinding_point: pubkey(41),
+				blinded_hops: vec![
+					BlindedHop { blinded_node_id: pubkey(45), encrypted_payload: vec![0; 45] },
+					BlindedHop { blinded_node_id: pubkey(46), encrypted_payload: vec![0; 46] },
+				],
+			})
+			.build()
+			.unwrap();
+		if let Err(e) = offer.to_string().parse::<Offer>() {
+			panic!("error parsing offer: {:?}", e);
+		}
+
+		let mut builder = OfferBuilder::new("foo".into(), pubkey(42));
+		builder.offer.paths = Some(vec![]);
```

Comment:
I don't quite get why this should cause us to fail to parse

Reply:
I likely inferred it from this part of the spec:

```
  - if it is connected only by private channels:
    - MUST include `offer_paths` containing one or more paths to the node from
      publicly reachable nodes.
  - otherwise:
    - MAY include `offer_paths`.
```

Reply:
True... I guess it is more that we fail if given a `paths` TLV record with length 0. We are fine parsing it if the `paths` TLV record doesn't exist. Could remove this check if you prefer.

Reply:
Removed the check.